File: _c++/symulacja_organizmow/Swiat.cpp
//
//  Swiat.cpp
//  WirtualnySwiat
//
//  Created by Michal Tarnacki on 10/04/2022.
//

#include "WszystkieOrganizmy.h"
#include "ZmienneDoWczytania.hpp"
#include <SDL2/SDL.h>
#include <fstream>
#define MIEJSCE_ZAPISU "./save.game"


Swiat::Swiat(Plansza plansza, KonfiguracjaOrganizmow konfiguracja){
    SDL_Window * window=nullptr;
    SDL_Init(SDL_INIT_VIDEO);
    SDL_Surface* screen = nullptr;
    window = SDL_CreateWindow("Okno do sterowania z klawiatury",
                                  200, 200, 50,
                                  50, SDL_WINDOW_SHOWN);
    screen = SDL_GetWindowSurface(window);
    SDL_FillRect(screen, NULL, SDL_MapRGB(screen->format, 0xFF, 0xFF, 0xFF));
    this->plansza=plansza;
    if(konfiguracja.getLacznaLiczbaOrganizmow()>plansza.GetLiczbaPol()) throw "Za duzo zwierzat!!";
    WczytajOrganizmy(konfiguracja);
    LosujPozycje();
}
void Swiat::WykonajTure(){
        SortujOrganizmy();
        PrzywrocMozliwosciRozrodu();
    for (Organizm *organizm: organizmy){
        if(organizm->SprawdzCzyZyje()){
            organizm->Akcja();
            for (Organizm *organizm2: organizmy){
                if(organizm2->SprawdzCzyZyje()&&organizm->GetPolozenie()==organizm2->GetPolozenie()&&organizm!=organizm2){
                    organizm2->Kolizja(*organizm);
                }
            }
            organizm->SetPoprzedniePolozenie(organizm->GetPolozenie());
        }
    }
        UsunMartweOrganizmy();
        AktualizujNumerRundy();
}
void Swiat::RysujSwiat() const{
    bool found;
    Polozenie polozenie = Polozenie(0,0);
    std::cout << "+-----------------------------------------+" << std::endl;
    std::cout << "|Wirtualny Swiat - Michal Tarnacki s188627|" << std::endl;
    std::cout << "+-----------------------------------------+" << std::endl;
    for(int y=0; y<plansza.GetM()+2; y++){
        for(int x=0; x<plansza.GetN()+2; x++){
            if((x==0&&y==0)||(x==0&&y==plansza.GetM()+1)||(x==plansza.GetN()+1&&y==0)||(x==plansza.GetN()+1&&y==plansza.GetM()+1))
                std::cout << '+';
            else if (x==0||x==plansza.GetN()+1)
                std::cout << '|';
            else if (y==0||y==plansza.GetM()+1)
                std::cout << '-';
            else{
                polozenie.setXY(x-1, y-1);
                found=false;
                for (Organizm *organizm: organizmy){
                    if(organizm->SprawdzCzyZyje()&&organizm->GetPolozenie()==polozenie){
                        std::cout << organizm->Rysowanie();
                        found = true;
                    }
                }
                if(!found)
                    std::cout << ' ';
            }
        }
        std::cout << std::endl;
    }
    std::cout << "Aktualna liczba organizmow:" << liczbaOrganizmow << std::endl;
}
void Swiat::Symuluj(){
    SterujCzlowiekiem(Polnoc);
    SDL_Event event;
    bool quit=false;
    bool wczytaj=false;
    WyczyscEkran();
    std::cout << "Wcisnij W aby wczytac poprzedni stan lub dowolny klawisz aby zaczac od nowa";
    while(!quit){
        while( SDL_PollEvent( &event ) ){
            switch( event.type ){
                case SDL_KEYDOWN:{
                    if(event.key.keysym.sym==SDLK_w){
                            wczytaj=true;
                            quit=true;
                        }
                    else
                        quit=true;
            }
            break;
          }
        }
    }
    if(wczytaj)
        Wczytaj();
    WyczyscEkran();
    RysujSwiat();
    while(!SprawdzCzyZostalTylkoJedenGatunek()){
        quit=false;
        while(!quit){
            while( SDL_PollEvent( &event ) ){
                switch( event.type ){
                    case SDL_KEYDOWN:{
                        if(event.key.keysym.sym==SDLK_UP){
                            SterujCzlowiekiem(Polnoc);
                            quit=true;
                            }
                        else if(event.key.keysym.sym==SDLK_RIGHT){
                            SterujCzlowiekiem(Wschod);
                            quit=true;
                        }
                        else if(event.key.keysym.sym==SDLK_DOWN){
                            SterujCzlowiekiem(Polodnie);
                            quit=true;
                        }
                        else if(event.key.keysym.sym==SDLK_LEFT){
                            SterujCzlowiekiem(Zachod);
                            quit=true;
                        }
                        else if(event.key.keysym.sym==SDLK_r){
                            if(!IstniejeCzlowiek())
                                quit=true;
                            else
                                WlaczSpecjalnaUmiejetnosc();
                        }
                        else if(event.key.keysym.sym==SDLK_s){
                            ZapiszAktualnyStan();
                            std::cout << "Zapisano aktualny stan:" << std::endl;
                        }
                        else{
                            if(!IstniejeCzlowiek())
                                quit=true;
                        }
                }
                break;
              }
            }
        }
        WyczyscEkran();
        WykonajTure();
        RysujSwiat();
    }
    remove(MIEJSCE_ZAPISU);
}
void Swiat::SortujOrganizmy(){
    std::list<Organizm*>::iterator ptrI;
    std::list<Organizm*>::iterator ptrJ;
    Organizm* organizmTmp;
    Organizm* organizmTmp2;
    for(int i=0; i<organizmy.size()-1; i++){
        ptrJ = organizmy.begin();
        ptrI = organizmy.begin();
        std::advance(ptrI, 1);
        for(int j=0; j<organizmy.size()-i-1; j++){
            organizmTmp=*ptrJ;
            organizmTmp2=*ptrI;
            if(organizmTmp->GetInicjatywa()<organizmTmp2->GetInicjatywa()||((organizmTmp->GetInicjatywa()==organizmTmp2->GetInicjatywa())&&organizmTmp->GetDlugoscZycia()<organizmTmp2->GetDlugoscZycia()))
                std::swap(*ptrI, *ptrJ);
            if(j==organizmy.size()-2){
                ptrJ = organizmy.begin();
                ptrI = organizmy.begin();
                std::advance(ptrI, 1);
            }
            else{
                std::advance(ptrJ, 1);
                std::advance(ptrI, 1);
            }
        }
    }
}

void Swiat::WczytajOrganizmy(KonfiguracjaOrganizmow konfiguracja){
    Organizm * nowyOrganizm;
    for(int i=0; i<konfiguracja.liczbaLudzi; i++){
        nowyOrganizm = new Czlowiek(*this);
        organizmy.push_back(nowyOrganizm);
    }
    for(int i=0; i<konfiguracja.liczbaAntylop; i++){
        nowyOrganizm = new Antylopa(*this);
        organizmy.push_back(nowyOrganizm);
    }
    for(int i=0; i<konfiguracja.liczbaWilkow; i++){
        nowyOrganizm = new Wilk(*this);
        organizmy.push_back(nowyOrganizm);
    }
    for(int i=0; i<konfiguracja.liczbaOwiec; i++){
        nowyOrganizm = new Owca(*this);
        organizmy.push_back(nowyOrganizm);
    }
    for(int i=0; i<konfiguracja.liczbaLisow; i++){
        nowyOrganizm = new Lis(*this);
        organizmy.push_back(nowyOrganizm);
    }
    for(int i=0; i<konfiguracja.liczbaZolwi; i++){
        nowyOrganizm = new Zolw(*this);
        organizmy.push_back(nowyOrganizm);
    }
    for(int i=0; i<konfiguracja.liczbaCyberOwiec; i++){
        nowyOrganizm = new CyberOwca(*this);
        organizmy.push_back(nowyOrganizm);
    }
    for(int i=0; i<konfiguracja.liczbaTraw; i++){
        nowyOrganizm = new Trawa(*this);
        organizmy.push_back(nowyOrganizm);
    }
    for(int i=0; i<konfiguracja.liczbaMleczy; i++){
        nowyOrganizm = new Mlecz(*this);
        organizmy.push_back(nowyOrganizm);
    }
    for(int i=0; i<konfiguracja.liczbaGuarany; i++){
        nowyOrganizm = new Guarana(*this);
        organizmy.push_back(nowyOrganizm);
    }
    for(int i=0; i<konfiguracja.liczbaWilczychJagod; i++){
        nowyOrganizm = new WilczeJagody(*this);
        organizmy.push_back(nowyOrganizm);
    }
    for(int i=0; i<konfiguracja.liczbaBarszuSosnowskiego; i++){
        nowyOrganizm = new BarszczSosnowskiego(*this);
        organizmy.push_back(nowyOrganizm);
    }
    liczbaOrganizmow=konfiguracja.getLacznaLiczbaOrganizmow();
}

void Swiat::LosujPozycje(){
    Polozenie polozenie;
    bool uniq;
    for (Organizm *organizm: organizmy){
        do{
        polozenie.setXY(rand()%plansza.GetN(),rand()%plansza.GetM());
            uniq=true;
            for (Organizm *organizm2: organizmy){
                if(polozenie==organizm2->GetPolozenie())
                    uniq=false;
            }
        } while(!uniq);
        organizm->SetPolozenie(polozenie);
        organizm->SetPoprzedniePolozenie(polozenie);
    }
};

Plansza Swiat::GetPlansza(){
    return this->plansza;
}

void Swiat::ZabijOrganizm(Organizm& organizm){
    organizm.Zabij();
    liczbaOrganizmow--;
};
void Swiat::NarodzOrganizm(Organizm& organizm, Polozenie polozenie){
    Organizm* nowyOrganizm = organizm.Urodz();
    nowyOrganizm->SetPolozenie(polozenie);
    nowyOrganizm->SetPoprzedniePolozenie(polozenie);
    nowyOrganizm->SetZdolnoscRozrodu(false);
    organizmy.push_front(nowyOrganizm);
    liczbaOrganizmow++;
};

void Swiat::UsunMartweOrganizmy(){
    Organizm* organizm=nullptr;
    for (Organizm *organizm2: organizmy){
        if(organizm!=nullptr){
            organizmy.remove(organizm);
            delete organizm;
            organizm=nullptr;
        }
        if(!organizm2->SprawdzCzyZyje())
            organizm=organizm2;
    }
    organizm=organizmy.back();
    if(!organizm->SprawdzCzyZyje()){
        organizmy.remove(organizm);
        delete organizm;
    }
}
std::list <Organizm*>& Swiat::GetOrganizmy(){
    return organizmy;
};

void Swiat::PrzywrocMozliwosciRozrodu(){
    for (Organizm *organizm: organizmy){
        if(!organizm->CzyMozeSieRozmnazac())
            organizm->SetZdolnoscRozrodu(true);
    };
}
void Swiat::AktualizujNumerRundy(){
    numerRundy++;
};
int Swiat::GetNumerRundy(){
    return numerRundy;
};
bool Swiat::SprawdzCzyZostalTylkoJedenGatunek() const {
    if(liczbaOrganizmow!=plansza.GetLiczbaPol())
        return false;
    Organizm* organizm = organizmy.front();
    for(Organizm* organizm2: organizmy){
        if(!organizm2->MozeBycPartnerem(*organizm)&&organizm2->SprawdzCzyZyje())
            return false;
    }
    return true;
};

void Swiat::ZapiszAktualnyStan(){
    std::fstream plik;
    plik.open(MIEJSCE_ZAPISU, std::ios::out);
    if(plik.is_open()==true){
        plik <<= plansza;
        plik << " " << numerRundy << " " << liczbaOrganizmow << std::endl;
        for(Organizm* organizm:organizmy){
            plik <<= *organizm;
            plik << std::endl;
        }
        plik.close();
    }
    else{
        std::cout << "nie udalo sie otworzyc pliku";
    }
};
void Swiat::Wczytaj(){
    std::fstream plik;
    plik.open(MIEJSCE_ZAPISU, std::ios::in);
    if(plik.is_open()==true){
        ZniszczSwiat();
        ZmienneDoWczytania zmienne(*this);
        plik >> zmienne.y >> zmienne.x >> zmienne.numerRundy >> zmienne.liczbaOrganizmow;
        this->plansza.set(zmienne.x,zmienne.y);
        this->numerRundy=zmienne.numerRundy;
        this->liczbaOrganizmow=zmienne.liczbaOrganizmow;
        for(int i=0; i<this->liczbaOrganizmow; i++){
            plik >> zmienne.znak;
            for(Organizm* organizm: zmienne.tymczasoweOrganizmy){
                if(zmienne.znak==organizm->Rysowanie()){
                    organizmy.push_back(organizm->Urodz());
                    plik >> zmienne.czasUrodzenia >> zmienne.sila >> zmienne.inicjatywa >> zmienne.x >> zmienne.y >> zmienne.x2 >> zmienne.y2 >> zmienne.czyZywy >> zmienne.zdolnyDoRozrodu;
                    if(dynamic_cast<Czlowiek*>(organizm))
                        plik >> zmienne.parametr1 >> zmienne.parametr2;
                    zmienne.polozenie.setXY(zmienne.x, zmienne.y);
                    zmienne.poprzedniePolozenie.setXY(zmienne.x2, zmienne.y2);
                    organizmy.back()->WczytajWlasciwosci(zmienne.sila, zmienne.inicjatywa, zmienne.polozenie, zmienne.poprzedniePolozenie, zmienne.czasUrodzenia, zmienne.czyZywy, zmienne.zdolnyDoRozrodu, zmienne.parametr1, zmienne.parametr2);
                    break;
                }
            }
        }
        plik.close();
    }
    else{
        std::cout << "nie udalo sie otworzyc pliku";
    }
}

Swiat::~Swiat(){
    ZniszczSwiat();
};
void Swiat::ZniszczSwiat(){
    for(Organizm *organizm: organizmy){
        organizm->Zabij();
        liczbaOrganizmow--;
    }
    UsunMartweOrganizmy();
}

void Swiat::WyczyscEkran() const{
    for(int i=0; i<100; i++) std::cout << std::endl;
//    system("CLEAR");
}

void Swiat::SterujCzlowiekiem(KierunekRuchu kierunek){
    Czlowiek *org;
    for(Organizm *organizm: organizmy){
        if(dynamic_cast<Czlowiek*>(organizm)){
            org=dynamic_cast<Czlowiek*>(organizm);
            org->UstawKierunekRuchu(kierunek);
        }
    }
}
void Swiat::WlaczSpecjalnaUmiejetnosc(){
    Czlowiek *org;
    for(Organizm *organizm: organizmy){
        if(dynamic_cast<Czlowiek*>(organizm)){
            org=dynamic_cast<Czlowiek*>(organizm);
            org->SpecjalnaUmiejetnosc();
        }
    }
}
bool Swiat::IstniejeCzlowiek(){
    for(Organizm *organizm: organizmy){
        if(dynamic_cast<Czlowiek*>(organizm)){
            return true;
        }
    }
    return false;
}

File: _java/symulacja_organizmow/src/Symulacja/Swiat.java
package Symulacja;

import java.io.*;
import java.util.*;
import javax.swing.*;
import Symulacja.Sluchacz;

public class Swiat {
    protected Plansza plansza = new Plansza();
    protected int liczbaOrganizmow;
    protected List<Organizm> organizmy = new ArrayList<>();
    protected List<Organizm> noweOrganizmy = new ArrayList<>();
    public Sluchacz zdarzenia = new Sluchacz();
    protected int numerRundy = 0;
    static final String miejsceZapisu = "plik.txt";
    protected Okno okno;
    protected KierunekRuchu kierunekCzlowieka = KierunekRuchu.BRAK;

    public Swiat(Plansza plansza, KonfiguracjaOrganizmow konfiguracjaOrganizmow) {
        this.plansza.Set(plansza);
//        if (konfiguracjaOrganizmow.getLacznaLiczbaOrganizmow() > plansza.GetLiczbaPol()) //throw "Za duzo zwierzat!!";
        WczytajOrganizmy(konfiguracjaOrganizmow);
        LosujPozycje();
    }
    public Swiat() throws IOException {
//        if (konfiguracjaOrganizmow.getLacznaLiczbaOrganizmow() > plansza.GetLiczbaPol()) //throw "Za duzo zwierzat!!";
//        Wczytaj();
    }
    public void ZainicjujOkno(){
        okno = new Okno("Symulacja", this);
    }

    protected void WykonajTure() throws FileNotFoundException, UnsupportedEncodingException, InterruptedException {
        SortujOrganizmy();
        PrzywrocMozliwosciRozrodu();
        okno.Update();
        Iterator<Organizm> iterator=organizmy.iterator();
        int size=organizmy.size();
        Organizm organizm;
        Organizm organizm2;
        for (int i=0; i<size; i++){
            organizm=organizmy.get(i);
            if(organizm.SprawdzCzyZyje()){
                organizm.Akcja();
                int size2=organizmy.size();
                for (int j=0; j<size2; j++){
                    organizm2=organizmy.get(j);
                    if(organizm2.SprawdzCzyZyje()&&organizm.GetPolozenie().equals(organizm2.GetPolozenie())&&organizm!=organizm2){
                        organizm2.Kolizja(organizm);
                    }
                }
                organizm.SetPoprzedniePolozenie(organizm.GetPolozenie());
            }
        }
        UsunMartweOrganizmy();
        AktualizujNumerRundy();
    }

    public void Symuluj() throws IOException, InterruptedException {
        boolean quit = false;
        boolean wczytaj = false;
        okno.setSize(800, 600);
        okno.setVisible(true);
        while(!SprawdzCzyZostalTylkoJedenGatunek()){
//            okno.Update();
            while (okno.buttonNum==0) {
                try {
                    Thread.sleep(10);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
                if(okno.buttonNum==2){
                    Wczytaj();
                    zdarzenia.Wyczysc();
                    okno.Update();
                    okno.buttonNum=0;
                }
                else if(okno.buttonNum==3){
                    ZapiszAktualnyStan();
                    okno.Update();
                    okno.buttonNum=0;
                }
                else if(okno.buttonNum==4){
                    SterujCzlowiekiem(KierunekRuchu.Polnoc);
                    okno.Update();
                    okno.buttonNum=0;
                }
                else if(okno.buttonNum==5){
                    SterujCzlowiekiem(KierunekRuchu.Zachod);
                    okno.Update();
                    okno.buttonNum=0;
                }
                else if(okno.buttonNum==6){
                    SterujCzlowiekiem(KierunekRuchu.Polodnie);
                    okno.Update();
                    okno.buttonNum=0;
                }
                else if(okno.buttonNum==7){
                    SterujCzlowiekiem(KierunekRuchu.Wschod);
                    okno.Update();
                    okno.buttonNum=0;
                }
                else if(okno.buttonNum==8){
                    WlaczSpecjalnaUmiejetnosc();
                    okno.Update();
                    okno.buttonNum=0;
                }
                else {
                    continue;
                }
            }
            zdarzenia.Wyczysc();
            okno.buttonNum=0;
            quit=true;
            WykonajTure();
            okno.Update();
            System.gc();
        }
    }

    private void RysujSwiat() {
        boolean found;
        Polozenie polozenie = new Polozenie(0, 0);
        System.out.println("+-----------------------------------------+");
        System.out.println("|Wirtualny Symulacja.Swiat - Michal Tarnacki s188627|");
        System.out.println("+-----------------------------------------+");
        for (int y = 0; y < plansza.GetM() + 2; y++) {
            for (int x = 0; x < plansza.GetN() + 2; x++) {
                if ((x == 0 && y == 0) || (x == 0 && y == plansza.GetM() + 1) || (x == plansza.GetN() + 1 && y == 0) || (x == plansza.GetN() + 1 && y == plansza.GetM() + 1))
                    System.out.print('+');
                else if (x == 0 || x == plansza.GetN() + 1)
                    System.out.print('|');
                else if (y == 0 || y == plansza.GetM() + 1)
                    System.out.print('-');
                else {
                    polozenie.Set(x-1, y-1);
                    found=false;
                    for (Symulacja.Organizm organizm: organizmy){
                        if(organizm.SprawdzCzyZyje()&&organizm.GetPolozenie().equals(polozenie)){
                            System.out.print(organizm.Rysowanie());
                            found = true;
                        }
                    }
                    if(!found)
                    System.out.print(' ');
                }
            }
            System.out.print("\n");
        }
        System.out.println("Aktualna liczba organizmow:" + liczbaOrganizmow);
    }

    protected void SortujOrganizmy() {
        organizmy.sort(Organizm::compareTo);
    }

    protected void WczytajOrganizmy(KonfiguracjaOrganizmow konfiguracja) {
        Symulacja.Organizm nowyOrganizm;
        for (int i = 0; i < konfiguracja.liczbaLudzi; i++) {
            nowyOrganizm = new Symulacja.Czlowiek(this);
            organizmy.add(nowyOrganizm);
        }
        for (int i = 0; i < konfiguracja.liczbaAntylop; i++) {
            nowyOrganizm = new Symulacja.Zwierzeta.Antylopa(this);
            organizmy.add(nowyOrganizm);
        }
        for (int i = 0; i < konfiguracja.liczbaWilkow; i++) {
            nowyOrganizm = new Symulacja.Zwierzeta.Wilk(this);
            organizmy.add(nowyOrganizm);
        }
        for (int i = 0; i < konfiguracja.liczbaOwiec; i++) {
            nowyOrganizm = new Symulacja.Zwierzeta.Owca(this);
            organizmy.add(nowyOrganizm);
        }
        for (int i = 0; i < konfiguracja.liczbaLisow; i++) {
            nowyOrganizm = new Symulacja.Zwierzeta.Lis(this);
            organizmy.add(nowyOrganizm);
        }
        for (int i = 0; i < konfiguracja.liczbaZolwi; i++) {
            nowyOrganizm = new Symulacja.Zwierzeta.Zolw(this);
            organizmy.add(nowyOrganizm);
        }
        for (int i = 0; i < konfiguracja.liczbaCyberOwiec; i++) {
            nowyOrganizm = new Symulacja.Zwierzeta.CyberOwca(this);
            organizmy.add(nowyOrganizm);
        }
        for (int i = 0; i < konfiguracja.liczbaTraw; i++) {
            nowyOrganizm = new Symulacja.Rosliny.Trawa(this);
            organizmy.add(nowyOrganizm);
        }
        for (int i = 0; i < konfiguracja.liczbaMleczy; i++) {
            nowyOrganizm = new Symulacja.Rosliny.Mlecz(this);
            organizmy.add(nowyOrganizm);
        }
        for (int i = 0; i < konfiguracja.liczbaGuarany; i++) {
            nowyOrganizm = new Symulacja.Rosliny.Guarana(this);
            organizmy.add(nowyOrganizm);
        }
        for (int i = 0; i < konfiguracja.liczbaWilczychJagod; i++) {
            nowyOrganizm = new Symulacja.Rosliny.WilczeJagody(this);
            organizmy.add(nowyOrganizm);
        }
        for (int i = 0; i < konfiguracja.liczbaBarszuSosnowskiego; i++) {
            nowyOrganizm = new Symulacja.Rosliny.BarszczSosnowskiego(this);
            organizmy.add(nowyOrganizm);
        }
        liczbaOrganizmow = konfiguracja.getLacznaLiczbaOrganizmow();
    }

    protected void LosujPozycje() {
        Polozenie polozenie = new Polozenie();
        boolean uniq;
        for (Organizm organizm : organizmy) {
            do {
                polozenie.Set((int) (Math.random()*plansza.GetN()), (int) (Math.random()*plansza.GetM()));
                uniq = true;
                for (Organizm organizm2 : organizmy) {
                    if (polozenie.equals(organizm2.GetPolozenie()))
                        uniq = false;
                }
            } while (!uniq);
            organizm.SetPolozenie(polozenie);
            organizm.SetPoprzedniePolozenie(polozenie);
        }
    };

    public Plansza GetPlansza() {
        return new Plansza(this.plansza);
    }

    private void KolejnoscRuchu() {
    }

    public void ZabijOrganizm(Organizm organizm) {
        organizm.Zabij();
        liczbaOrganizmow--;
    }

    ;

    public void NarodzOrganizm(Organizm organizm, Polozenie polozenie, TrybNarodz tryb) {
        Organizm nowyOrganizm = organizm.Urodz();
        nowyOrganizm.SetPolozenie(polozenie); //2 3 // 3 1 +1 -2
        nowyOrganizm.SetPoprzedniePolozenie(polozenie);
        nowyOrganizm.SetZdolnoscRozrodu(false);
        organizmy.add(nowyOrganizm);
        liczbaOrganizmow++;
        System.out.println("Rodzi sie na " + polozenie);
        if(tryb == TrybNarodz.Dodaj)
          okno.Update();
    }

    protected void UsunMartweOrganizmy() {
        Organizm organizm=null;
        noweOrganizmy.clear();
        for (Organizm organizm2: organizmy){
            if(organizm2.SprawdzCzyZyje())
                noweOrganizmy.add(organizm2);
        }
        organizmy.clear();
        organizmy.addAll(noweOrganizmy);
    }
    ;

    public List<Organizm> GetOrganizmy() {
        return organizmy;
    };

    protected void PrzywrocMozliwosciRozrodu() {
        for (Organizm organizm : organizmy) {
            if (!organizm.CzyMozeSieRozmnazac())
                organizm.SetZdolnoscRozrodu(true);
        }
    }

    protected void AktualizujNumerRundy() {
        numerRundy++;
    };

    public int GetNumerRundy() {
        return numerRundy;
    };

    protected boolean SprawdzCzyZostalTylkoJedenGatunek() {
        if (liczbaOrganizmow != plansza.GetLiczbaPol())
            return false;
        Organizm organizm = organizmy.get(0);
        for (Organizm organizm2:organizmy){
            if (!organizm2.MozeBycPartnerem(organizm)&&organizm2.SprawdzCzyZyje())
            return false;
        }
        return true;
    }

    ;

    protected void ZapiszAktualnyStan() throws FileNotFoundException, UnsupportedEncodingException {
        PrintWriter writer = new PrintWriter(miejsceZapisu, "UTF-8");
        writer.println(plansza.toString(TrybToString.Krotko));
        writer.println(numerRundy + " " + liczbaOrganizmow);
            for (Organizm organizm:organizmy){
                writer.println(organizm.toString(TrybToString.Krotko));
            }
        writer.close();
    }

    ;
    protected void Wczytaj() throws IOException {
        File plik = new File(miejsceZapisu);
        if(!plik.exists())
            return;
        Scanner reader = new Scanner(plik);
            ZniszczSwiat();
            ZmienneDoWczytania zmienne = new ZmienneDoWczytania(this);
            zmienne.x = reader.nextInt();
            zmienne.y = reader.nextInt();
            zmienne.numerRundy = reader.nextInt();
            zmienne.liczbaOrganizmow = reader.nextInt();;
            this.plansza.Set(zmienne.x,zmienne.y);
            this.numerRundy=zmienne.numerRundy;
            this.liczbaOrganizmow=zmienne.liczbaOrganizmow;
            for(int i=0; i<this.liczbaOrganizmow; i++){
                zmienne.znak = reader.next().charAt(0);;
                for(Organizm organizm: zmienne.tymczasoweOrganizmy){
                    if(zmienne.znak==organizm.Rysowanie()){
                        organizmy.add(organizm.Urodz());
                        zmienne.czasUrodzenia = new Date(reader.nextLong());
                        zmienne.sila = reader.nextInt();
                        zmienne.inicjatywa = reader.nextInt();
                        zmienne.x = reader.nextInt();
                        zmienne.y = reader.nextInt();
                        zmienne.x2 = reader.nextInt();
                        zmienne.y2 = reader.nextInt();
                        zmienne.czyZywy = reader.nextInt()!=0 ? true:false;
                        zmienne.zdolnyDoRozrodu = reader.nextInt()!=0 ? true:false;
                        if(organizm instanceof Czlowiek) {
                           zmienne.parametr1 = reader.nextInt()!=0 ? true:false;
                            zmienne.parametr2 = reader.nextInt();
                        }
                        zmienne.polozenie.Set(zmienne.x, zmienne.y);
                        zmienne.poprzedniePolozenie.Set(zmienne.x2, zmienne.y2);
                        organizmy.get(organizmy.size()-1).WczytajWlasciwosci(zmienne.sila, zmienne.inicjatywa, zmienne.polozenie, zmienne.poprzedniePolozenie, zmienne.czasUrodzenia, zmienne.czyZywy, zmienne.zdolnyDoRozrodu, zmienne.parametr1, zmienne.parametr2);
                        break;
                    }
                }
            }
            organizmy.get(0);
            reader.close();
    }

    protected void ZniszczSwiat() {
        for (Organizm organizm : organizmy) {
            organizm.Zabij();
            liczbaOrganizmow--;
        }
        UsunMartweOrganizmy();
    }

    public void SterujCzlowiekiem(Symulacja.KierunekRuchu kierunek) {
        Symulacja.Czlowiek org;
        kierunekCzlowieka=kierunek;
        for (Symulacja.Organizm organizm : organizmy) {
            if (organizm instanceof Symulacja.Czlowiek) {
                org = (Symulacja.Czlowiek) organizm;
                org.UstawKierunekRuchu(kierunek);
            }
        }
    }
    protected String KierunekCzlowiekaToString(){
        if(kierunekCzlowieka==KierunekRuchu.Polnoc)
            return "Gora";
        else if(kierunekCzlowieka==KierunekRuchu.Polodnie)
            return "Dol";
        else if(kierunekCzlowieka==KierunekRuchu.Zachod)
            return "Lewo";
        else if(kierunekCzlowieka==KierunekRuchu.Wschod)
            return "Prawo";
        else
            return "brak";

    }

    public void WlaczSpecjalnaUmiejetnosc() {
        Czlowiek org;
        for (Organizm organizm : organizmy) {
            if (organizm instanceof Symulacja.Czlowiek) {
                org = (Symulacja.Czlowiek) organizm;
                org.SpecjalnaUmiejetnosc();
            }
        }
    }

    public boolean IstniejeCzlowiek() {
        for (Organizm organizm : organizmy) {
            if (organizm instanceof Symulacja.Czlowiek) {
                return true;
            }
        }
        return false;
    }
};

